### DISCUSSION TITLE
Mark links?

### DISCUSSION THREAD

**Discussion by groves** [kovidgoyal/kitty#5442] [Q&A]
I have [this kitten](https://github.com/groves/catherd/blob/main/linky.py) that lets me step through the OSC 8 hyperlinks from the last command, editing each one in turn. With hyperlinked_grep, it's kinda like doing a "Find" in a browser page and then hitting the "Find Next" keyboard shortcut to go through the matches.

Like with finding in a browser, I'd like to highlight all the matches with one color and to highlight the currently visited match with a second color. Makes it easier to keep context.

Kitty's marks seem ideal for this, but I don't see a way to get ANSI input with OSC 8 hyperlinks in it to a marker function. Without the hyperlinks, it's hard to know what to mark.

Even if I had hyperlinks, if there were identical hyperlinks, I wouldn't know which one to highlight as current since the marker only has the current line. If it were able to emit ranges over the entire input like the custom processor for hints, that would let me mark the correct link. I don't think there's a way to call a marker over a whole region though?

If there isn't, I'd be down to submit a PR to add the ability to create marker functions that mark a whole text range, not just a single line. I wanted to check that there wasn't already a way to do this before starting on that though.

**Comment by kovidgoyal** [comment]
There isn't scanning is deliberately limited to a line for performance
reasons, as marks are evaluated in the hot path, during input processing.

Your use case would be better served by a command line kitten/program
that gets the last command output and parses it however it likes and
responds to keypresses and does whatever else you like. The only missing
piece in the existing kitty infrastructure is getting the commandline of
the last executed command. The best way to do that would be to use an
escape code like the one for last reported cwd (OSC 7 IIRC) and have the
shell send it to kitty via the shell integration scripts. THe kitten can
then access that information.

That should be a fully robust and flexible solution.

**Reply by groves** [reply]
> scanning is deliberately limited to a line for performance reasons, as marks are evaluated in the hot path, during input processing.

That's helpful context. Thanks!

> a command line kitten/program that gets the last command output and parses it however it likes

That seems like it would need to implement its own scrolling and history management and wouldn't allow the terminal to be used normally while it's active.

I'm trying to figure out a way to write tools like this collaboratively with the existing terminal UI. That would let you layer in functionality, which has two benefits. Authors of tools wouldn't need to recreate all the display/scroll/click and other handling that the terminal/shell is already doing, and users of the tools wouldn't be taken out of their shells into a new application with a different interaction paradigm.

Maybe layering on top of the terminal UI is a pipe dream, but it feels like there's enough benefit to try to figure it out.

What if we optionally gave a line number to mark functions and as_text callers. It seems like there should be a monotonically increasing line number for output into the terminal, where the content for a line number is fixed once it transitions off the screen into the history buffer. While it's on the screen, you could still get a line number for given text, the text might just change.

I know some of those line numbers would disappear as they rolled out of the history buffer, but it still seems like they could function as an identifier for marking like this? I could build a dict of line numbers to marks when I process input, and then use that dict to decide what to mark in the marker function.

**Comment by kovidgoyal** [comment]
On Mon, Aug 29, 2022 at 01:01:43PM -0700, Charlie Groves wrote:
> > scanning is deliberately limited to a line for performance reasons, as marks are evaluated in the hot path, during input processing.
> 
> That's helpful context. Thanks!
> 
> > a command line kitten/program that gets the last command output and parses it however it likes
> 
> That seems like it would need to implement its own scrolling and history management and wouldn't allow the terminal to be used normally while it's active.

Not really, you would need to implement it once and then re-use it for
all such tools. It would actually be less effort than trying to shoehorn
it into kitty. Or you could ask the author of less to have a keybinding
for jump to next hyperlink, and just use less. Which is what kitty uses
for most of its UI anyway. Scrollback, --help output, etc.

> 
> I'm trying to figure out a way to write tools like this collaboratively with the existing terminal UI. That would let you layer in functionality, which has two benefits. Authors of tools wouldn't need to recreate all the display/scroll/click and other handling that the terminal/shell is already doing, and users of the tools wouldn't be taken out of their shells into a new application with a different interaction paradigm.

The idea is to re-use this code in a pager like a less not re-write it
each time.

> 
> Maybe layering on top of the terminal UI is a pipe dream, but it feels like there's enough benefit to try to figure it out.

If you want jump to hyperlink functionality in kitty, just create an
action for it, should be pretty trivial to implement and doesnt have
performance implications. You can have it scan the buffer for hyperlink
or pattern of interest and scroll to it. Even have an action to trigger
last scrolled to hyperlink.

**Reply by kovidgoyal** [reply]
Indeed thinking about it you could even have the action highlight the cells corresponding to the hyperlink temporarily. 

So I think the way this should work if you want it integrated in kitty is create an action that triggers a "jump to matched content mode". This mode will be like the existing select_window mode implemented in boss.py in that it will have dedicated keybindings once triggerred. 

The action should take arguments controlling what to match which can be a regex that matches text or a hyperlink. It should scan the history and screen buffers and scroll to the first match when triggerred, highlighting it. Then it can jump to next/previous using some keys like n/p whatever.  and when pressing enter it can trigger the hyperlink (or for text matches either copy to clipboard or insert into window). pressing esc should clear the mode.

While the mode is active you can highlight the matched cells with an underline or similar style and restore them when the mode is canceled.

**Reply by groves** [reply]
> The idea is to re-use this code in a pager like a less not re-write it each time. 

That makes a lot more sense than my interpretation. I'm so used to a pager being extremely minimal that I didn't think to imagine what one with some modern terminal handling could do. You could do this sort of navigation, correctly insert the paged contents into the scrollback, use OSC 52 to copy from the pager into the clipboard, and likely lots more. Thanks for clarifying!

I'm not even sure that the overlay I pushed for that you described adds much at that point. Like you said, once the pager isn't anemic, you can draw whatever you want. That would let the pager keep the existing styling of the displayed text while drawing its highlighting and navigation on top. I can't think of anything that would be necessary at the kitty level at that point. Will have to play with it though...

**Reply by kovidgoyal** [reply]
Yes, indeed, it is actually on my TODO list to create a pager to replace less that has modern capabilities. Sadly my TODO list stretches to the moon, so it will be some time before I can get to it.

**Reply by groves** [reply]
>  it is actually on my TODO list to create a pager to replace less that has modern capabilities

I ended up writing a pager that lets me navigate to hyperlinks: https://github.com/groves/ate It lets me do the navigation between search results I was after here.

I'm using launch to integrate it into kitty:

`map F8 launch --type=overlay --stdin-source @last_cmd_output --stdin-add-formatting --env ATE_OPENER=/Users/groves/.nix-profile/bin/edit ate`

With that, I can run hyperlinked_grep, hit F8, my pager gets launched, I can select whatever link I want, and hitting enter opens the match it in my editor. 

Did you have other pager features you wanted? Maybe I can add them to my pager and take creating one off your todo list 😎

**Comment by kovidgoyal** [comment]
On Wed, Nov 16, 2022 at 06:04:44PM -0800, Charlie Groves wrote:
> >  it is actually on my TODO list to create a pager to replace less that has modern capabilities
> 
> I ended up writing a pager that lets me navigate to hyperlinks: https://github.com/groves/ate It lets me do the navigation between search results I was after here.

One note, from https://sevorg.org/posts/why_ate/ it seems you are using
the params field in OSC 8 to pass line/column data. I recommend using a
fragment in the URL instead, that way it will work with kitty's
open_actions which dont know anything about OSC 8 and work with URLs.
Something like:

file:///whatever#line[:col]

where the column number is optional.

> 
> I'm using launch to integrate it into kitty:
> 
> `map F8 launch --type=overlay --stdin-source @last_cmd_output --stdin-add-formatting --env ATE_OPENER=/Users/groves/.nix-profile/bin/edit ate`
> 
> With that, I can run hyperlinked_grep, hit F8, my pager gets launched, I can select whatever link I want, and hitting enter opens the match it in my editor. 
> 
> Did you have other pager features you wanted? Maybe I can add them to my pager and take creating one off your todo list 😎 

Well, my laundry list is quite long :)

In no particular order:

* Uses a standard based wcswidth() implementation to do wrapping
* Has a way to scroll to a particular line in the input when starting
* Support for copy/paste using vim visual mode like bindings
* Support for mouse interactions (scrolling with wheel/touchpad, dragging to select, drag scroll)
* Support for hyperlinks, in that clicking a hyperlink should open the clicked file in the pager and allow you to go back/forwards like a browser
* Support for user customizable actions when clicking hyperlinks
* Image support (stretch goal :)

**Reply by groves** [reply]
> One note, from https://sevorg.org/posts/why_ate/ it seems you are using the params field in OSC 8 to pass line/column data. I recommend using a fragment in the URL instead

That section is explaining OSC 8 itself, so the params there are to walk through how they work.

ate itself doesn't care about params vs anchors. It passes the whole content from OSC 8 into the opener. I'm using the anchor in the opener examples in the repo since it's easier to parse with a regex: https://github.com/groves/ate/blob/f68d33a4e49a1630ab20cacbfaae7a5e1487c3e9/opener_examples/nvim_opener.bash#L13

> Well, my laundry list is quite long :) In no particular order:
> * Uses a standard based wcswidth() implementation to do wrapping 

If I'm following you, I'm doing this already. I use [grapheme_column_width](https://docs.rs/termwiz/latest/termwiz/cell/fn.grapheme_column_width.html) to figure out how many cells a given grapheme takes up in the output. I then use those cell widths to figure out when to wrap.

Of course that relies on the Unicode versions matching between the text emitter, ate, and the terminal rendering it for perfect fidelity. Is there a good way to ask kitty what version of Unicode it's using to decide on grapheme widths? I could probably remove some differences between what ate's expecting and what kitty's rendering if I matched that unicode version.

> * Support for mouse interactions (scrolling with wheel/touchpad, dragging to select, drag scroll)

This is on my "soon" list.

> * Support for user customizable actions when clicking hyperlinks

I need to add the mouse support, but I believe this would be done by invoking ATE_OPENER with the clicked url. Since the opener is your own program, you can do whatever action you want in it.

Thanks for the list!

**Comment by kovidgoyal** [comment]
On Thu, Nov 17, 2022 at 04:13:26AM -0800, Charlie Groves wrote:
> If I'm following you, I'm doing this already. I use [grapheme_column_width](https://docs.rs/termwiz/latest/termwiz/cell/fn.grapheme_column_width.html) to figure out how many cells a given grapheme takes up in the output. I then use those cell widths to figure out when to wrap.
> 

Yes, that should be mostly fine.

> Of course that relies on the Unicode versions matching between the text emitter, ate, and the terminal rendering it for perfect fidelity. Is there a good way to ask kitty what version of Unicode it's using to decide on grapheme widths? I could probably remove some differences between what ate's expecting and what kitty's rendering if I matched that unicode version.

Unicode doesnt change all that much, its mostly newer symbol/emoji
these days. So in practice this is not an issue. If it does become
an issue I can always add it here: https://sw.kovidgoyal.net/kitty/kittens/query_terminal/

> 
> > * Support for mouse interactions (scrolling with wheel/touchpad, dragging to select, drag scroll)
> 
> This is on my "soon" list.
> 
> > * Support for user customizable actions when clicking hyperlinks
> 
> I need to add the mouse support, but I believe this would be done by invoking ATE_OPENER with the clicked url. Since the opener is your own program, you can do whatever action you want in it.

That wont work with back and forward though.
